Config post processing

weierophinney opened the pull request:
One common issue when using third-party libraries is that the configuration keys will still reference ZF artifacts: namespaces, classes, and configuration keys.

This patch provides a recursive configuration post-processor that rewrites those references using the rules used when performing migrations, ensuring that they now reference the Laminas equivalents.

The post processor, `Laminas\ZendFrameworkBridge\ConfigPostProcessor` can be used with laminas-config-aggregator. Additionally, you can register `Laminas\ZendFrameworkBridge` as a module in MVC applications, and it will register a `ModuleEvent::EVENT_CONFIG_MERGE` listener that will pass merged configuration to the post processor, and pass the transformed configuration back to the `ConfigListener`.

The module is also exposed via the `extra.laminas.module` directive, allowing end-users to opt-in to using the post processor in their MVC applications.

The migration tooling will be updated to register the module in MVC applications, and the post processor in Expressive applications.

This patch also introduces `Laminas\ZendFrameworkBridge\Replacements`, and it contains the default list of replacements, as pulled from the laminas-migration library. I suggest that we update laminas-migration to depend on this library, so that the replacements are in a single location.

Finally, I've updated the plugin to only support PHP 5.6 and up, for the same reasons the migration tooling and dependency plugin only support those versions and up.

### TODO

- [x] Fix how it handles dependency configuration in order to prevent cyclical aliasing.

weierophinney committed:
feat: config post processing

One common issue when using third-party libraries is that the
configuration keys will still reference ZF artifacts: namespaces,
classes, and configuration keys.

This patch provides a recursive configuration post-processor that
rewrites those references using the rules used when performing
migrations, ensuring that they now reference the Laminas equivalents.

The post processor, `Laminas\ZendFrameworkBridge\ConfigPostProcessor`
can be used with laminas-config-aggregator. Additionally, you can
register `Laminas\ZendFrameworkBridge` as a module in MVC applications,
and it will register a `ModuleEvent::EVENT_CONFIG_MERGE` listener that
will pass merged configuration to the post processor, and pass the
transformed configuration back to the `ConfigListener`.

The module is NOT exposed via the `extra.laminas.module` directive, to
ensure that it is opt-in, and thus can also be disabled.

The migration tooling will be updated to register the module in MVC
applications, and the post processor in Expressive applications.

I also recommend that we have the migrations package depend on this one,
so that all rewrite rules are in one package.

weierophinney committed:
qa: adds tests for Module class

Provides mocks for the various collaborators, and then tests
interactions.

weierophinney committed:
qa: import edge case tests from laminas-migration

This patch imports the edge case tests from laminas-migration.

weierophinney committed:
feat: expose as a module via Composer

weierophinney committed:
fix: do not merge non-array values

Updates the logic in ConfigPostProcessor to overwrite values if they
exist and are non-arrays, and only merge if they are both array values.

weierophinney committed:
fix: prevent cyclical alias definitions

The various Laminas packages all define aliases of the legacy ZF classes
to the Laminas counterparts. As such, if we were to rewrite those
aliases, we end up with a cyclical alias situation (e.g.,
`'Zend\Db\Adapter\Adapter' => 'Laminas\Db\Adapter\Adapter'` would become
`'Laminas\Db\Adapter\Adapter' => 'Laminas\Db\Adapter\Adapter'`).

As such, this patch modifies the `rewriteValue` functionality to take
into account the key being rewritten, and to skip rewriting for arrays
under the 'aliases' key.

weierophinney commented:
I've now tested this in a variety of full-application scenarios, and all worked perfectly following a migration. While it adds some initial overhead, that overhead is removed in production due to configuration caching.

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,107 @@
+            // Since we're rewriting at the PHP level, and not the raw
+            // strings from reading file contents, we need to add this
+            // one.
+            'zend-expressive' => 'expressive',
```
Is it safe even if any 3rd party library with name zend-expressive-* is using also zend-expressive-* configuration?
We are not rewriting that 3rd library source, so not sure if we can rewrite any configuration key in that case.

weierophinney replied:
That's a good question.

The problem is that we need to rewrite the config key `zend-expressive` to `expressive`.

However, there may _also_ be keys _starting_ with `zend-expressive` that would get rewritten in that case.

I can see the following solutions, and each poses... challenges:

- Leave as I've written it, and potentially have it rewrite keys we do not control.
- Add a mechanism for doing replacements on exact matches (pseudo-code: `exactMatchExists($key) ? exactReplace($key) : replacements($key)`. This could become a potential maintenance issue if we start identifying more such cases.

I think I'll go ahead with the second option (there aren't many, based on what I was identifying when we created the original lists for laminas-migration).

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,107 @@
+            return $this->replacements->replace($value);
+        }
+
+        if (null === $value || is_scalar($value) || is_object($value)) {
```
isn't it better to have just `! is_array($value)` condition here?

weierophinney replied:
Yes. /headdesk.

Thanks. :smile:

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,107 @@
+     * @param  bool  $preserveNumericKeys
+     * @return array
+     */
+    public static function merge(array $a, array $b, $preserveNumericKeys = false)
```
I know it's copy from zend-stdlib, but we are not using the 3rd param here, so I guess we can remove it.

weierophinney replied:
Good catch - thanks!

michalbundyra commented on src/Replacements.php:
```diff
@@ -0,0 +1,34 @@
+     */
+    public function replace($value)
+    {
+        if (! is_string($value)) {
```
why do we need that case here? we expecting that input parameter is string, and in PHP 7 we would add `string` typehint. Method shouldn't be called for anything what is not a string.

weierophinney replied:
I was being over-protective, in case somebody attempts to use the library elsewhere. :smile:

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,107 @@
+        foreach ($config as $key => $value) {
+            $newKey = is_string($key) ? $this->replacements->replace($key) : $key;
+
+            if (isset($rewritten[$newKey]) && is_array($rewritten[$newKey])) {
```
We are missing condition here that `$key !== $newKey`.

In case the key is exactly the same, and we are merging the contents we can duplicate values if arrays have numeric values. So I guess it will be bad for delegators:
```php
'delegators' => [
   Buzzer::class => [
        Delegator1Factory::class,
        Delegator2Factory::class,
   ],
],
```

the name `Buzzer::class` is gonna stay the same but in rewritten version we are gonna have:

```php
'delegators' => [
   Buzzer::class => [
        Delegator1Factory::class,
        Delegator2Factory::class,
        Delegator1Factory::class,
        Delegator2Factory::class,
   ],
],
```

unless - I am missing some condition somewhere?

michalbundyra replied:
oh... and here will be another issue:
```php
'delegators' => [
   Buzzer::class => [
        \MyPackage\Delegator1::class,
        \Zend\Package\Delegator2::class,
   ],
],
```

it is gonna be rewritten and merged into:

```php
'delegators' => [
   Buzzer::class => [
        \MyPackage\Delegator1::class,
        \Zend\Package\Delegator2::class,
        \MyPackage\Delegator1::class,
        \Laminas\Package\Delegator2::class,
   ],
],
```

weierophinney replied:
The purpose of this clause is to detect if the key already exists in the replacements array, and is itself an array value; if it does, it indicates they have a keys named after both a ZF class AND its equivalent Laminas class. It likely should not happen, but this catches it, and, in that scenario, will _merge_ both arrays.

In terms of integer keys, you should never have top-level integer keys anyways, and, if you _do_ have integer keys, you cannot have duplicates.

However, that `isset()` should likely be an `array_key_exists` to ensure there is no type-munging. I can make that change. I'm not going to add a check for `$key !== $newKey` unless I have a test case to test against, though.

weierophinney replied:
> unless - I am missing some condition somewhere?

You are - this is happening _after_ configuration merging, so we won't have two keys of the same name, unless there are references to both a string AND its replacement. There will never be duplicate integer keys.

michalbundyra replied:
Yes, you are correct.

So as discussed on Slack - 3rd party libraries can't really provide compatibility with ZF and Laminas/Expressive/Apigility packages. These must support ZF only (but because of our tools will be compatible with Laminas) or Laminas only, when our tools will be not applied.

weierophinney replied:
To summarize that discussion:

- 3rd party libraries get forwards compatibility for free when consumers use this post processor (as it will rewrite references to ZF classes)
- 3rd party libraries will eventually need to create a new major version that targets Laminas packages specifically, dropping support for the ZF versions.

michalbundyra replied:
@weierophinney yes, sound good 👍

michalbundyra commented on src/Module.php:
```diff
@@ -0,0 +1,56 @@
+class Module
+{
+    /** @var string[] */
+    private $replacements;
```
cannot see where it is used?

weierophinney replied:
Remnant of a previous implementation; I'll remove. Thanks!

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        // Array; time to recurse, but only if not aliases.
+        return $key === 'aliases' ? $value : $this($value);
```
we don't want to rewrite sub-keys of `aliases` arrays, but I think we should rewrite values, for example:

```php
'aliases' => [
    'Zend\Package\MyAlias' => 'Zend\Package\AwesomeClass',
],
```

we want to become:
```php
'aliases' => [
    'Zend\Package\MyAlias' => 'Laminas\Package\AwesomeClass',
],
```

it will reduce iterations in SM - without rewriting it should also work, unless... this very edge case but still possible:
```php
'factories' => [
    'Zend\Something' => Zend\...\InvokableFactory::class,
],
'aliases' => [
   'my-alias' => '\Zend\Something',
],
```

in current state we are gonna get:

```php
'factories' => [
    'Laminas\Something' => Laminas\...\InvokableFactory::class,
],
'aliases' => [
   'my-alias' => '\Zend\Something',
],
```

and calling `$sm->get('my-alias')` is gonna fail, as we do not have defined `Zend\Something` anymore.

weierophinney replied:
That's a great test case for me - I'll get to work on this shortly.

michalbundyra reviewed (changes requested):
@weierophinney Check out my comments, I think we can have couple edge cases we should also handle.

weierophinney committed:
fix: do not replace "zend-expressive" unilaterally

There may be third-party libraries that use "zend-expressive" as a
configuration key prefix. To avoid rewriting those, this patch adds
functionality for replacing exact matches, and then falls back to the
general rewrite rules (which omit that match).

weierophinney committed:
qa: simplify conditional

weierophinney committed:
qa: remove unused third argument to merge()

weierophinney committed:
qa: remove conditional type check

Ideally, we'd have a typehint here. But since it is documented as
accepting strings only, anybody using it with a non-string value should
likely receive a notice from PHP about it.

weierophinney committed:
qa: use more specific array_key_exists instead of isset

Prevents type munging.

weierophinney committed:
qa: remove obsolete property declaration

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,107 @@
+            $newKey = is_string($key) ? $this->replacements->replace($key) : $key;
+
+            if (isset($rewritten[$newKey]) && is_array($rewritten[$newKey])) {
+                $rewritten[$newKey] = self::merge($rewritten[$newKey], $this->rewriteValue($value, $newKey));
```
as discussed on slack - we need ensure that `$value` is array, not sure what we should do if one value is array and the other is not?

weierophinney replied:
Four options:

- Raise an exception
- Emit a notice
- Silently discard the old value (i.e., newer values take precedence over older ones)
- Use the following rules:
  - non-null values overwrite null values
  - array or object values overwrite scalar values
  - object values overwrite array values

And/or emit a notice AND do either the third or fourth option.

Not sure what users would expect here.

michalbundyra replied:
I think we need to check what we are doing now on merging configuration and take the same approach here. Currently we can have the situation that in zf library we have:
```php
'zend-expressive' => [ ... array configuration ... ]
```

and user configuration:
```php
'zend-expressive' => 'string',
```

weierophinney replied:
I'll check and see what zend-config-aggregator does, and do that. :smile:

weierophinney replied:
zend-config-aggregator uses the same rules as zend-stdlib `ArrayUtils::merge`, which has the following algorithm:

- If both values are arrays, they're merged.
- Otherwise, the new value overwrites the old value.

This works fine for the config aggregator, because it processes configuration in a determined order (config providers, then global config files, then local config files).

For the post processor, it could potentially be problematic, as it depends on the internal array order: if we rewrite something with a ZF key and already have configuration under the equivalent Laminas key, it could potentially mean overwriting something that was not expected to be overwritten.

That said: in most cases, these will be coming from third-party components, which means that the keys should occur _earlier_ in the array.

I'm going to update to follow this algorithm, and put a note in the README about the scenario.

weierophinney replied:
We already follow that algorithm! But I'm adding a test case as well, to ensure we continue to do so.

weierophinney committed:
fix: ensure we rewrite alias _targets_

This patch is actually a refactor of how replacements are made to use a
rules engine. Each rule is run, and can return null or a callable. If a
null is returned, it checks the next rule, and so-forth. If a callable
is returned, it is used to perform the replacement.

This approach allows us to have rules of different granularity, and for
specific conditions.

I rewrote the "exact replacements" functionality as a ruleset, and added
two more:

- one for when the key is "aliases" and it has an array value
- one for a non-null key and an array value

In the first case, it now invokes the new method
`replaceDependencyAliases`, which replaces the targets for aliases, but
not the keys. In the second case, it calls `__invoke()` on the value.

The fallback checks for a string value, returning the result of
`Replacements::replace()` on it if so, and the unchanged value
otherwise.

The `replaceValue()` method was renamed `replace`, and is now called
both for keys (in which case the `$key` value passed to a ruleset is
`null`) as well as values.

weierophinney committed:
fix: ensure we merge values correctly

Adds a test to ensure the precedence order when merging values. It
should be:

- If both values are arrays, they are merged, with the second taking
  precedence over the first.
- If the original value is not an array, the second value overwrites the
  original.

Additionally, this adds a check to see if the value provided to
`merge()` is an array; if not, it will append the value to the original
array.

weierophinney committed:
qa: remove lowest phpunit constraint

No longer relevant if we are targetting PHP 5.6 as the lowest supported
PHP version.

weierophinney commented:
I've pushed changes to address the various areas of feedback.

The code is already following the same merge algorithm as zend-config-aggregator, which means:

- If the first value is a scalar, the second value overwrites.
- If the first value is an array, and the second value is scalar, the second value is appended.
- If both values are arrays, they are merged, with the second taking precedence.

I've added tests to cover this as well, to make it explicit.

I have also refactored the internals to use _rulesets_. Each ruleset is a callable that receives the value and key, and then either returns a null or a callable. If a callable is returned, that value is used to perform replacements on the value. If no ruleset matches, the default ruleset is to replace string values using the string translation rules, and otherwise return the value verbatim (which was the previous behavior).

I have added the following rulesets:

- exact match: in this case, the value exactly matches something we know about, and we return the replacement.
- aliases: matched when the key is "aliases" and the value is an array. It will replace alias _targets_ only, but not the keys. This prevents circular aliasing issues.
- arrays: matched when a non-null key is present and the value is an array. It invokes itself on the value.

This approach should allow us to add more rulesets as we need them, while keeping the code maintainable (no switch/case or if/elseif madness).

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,194 @@
+{
+    /** @var array String keys => string values */
+    private $exactReplacements = [
+        'zend-expressive' => 'expressive',
```
We need here also (at least) `zf-apigility`.

weierophinney replied:
I thought we had no top-level key for that, but I just checked, and we do. I'll get that change in.

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,194 @@
+        $this->rulesets     = [
+            // Exact values
+            function ($value) {
+                return is_string($value) && array_key_exists($value, $this->exactReplacements)
```
minor: we can use here `isset` instead of `array_key_exists`. It has no sense replace string to `null`, and per definition `exactReplacements` property has string values only.

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    /**
+     * Perform subsitutions as needed on an individual value.
```
typo in `subsitutions`

michalbundyra commented on src/Replacements.php:
```diff
@@ -0,0 +1,31 @@
+    /** @var string[] */
+    private $replacements;
+
+    public function __construct(array $additionalReplacements = [])
```
we are not using this parameter now, so not sure if we should keep it

weierophinney replied:
I'd like to, as it provides an extension point if we want to expand the replacement list.

michalbundyra commented on test/TestAsset/ConfigPostProcessor/MergeEquivalentKeys.php:
```diff
@@ -0,0 +1,16 @@
+            'template_error' => 'custom::500',
+        ],
+    ],
+    'apigility-rest' => 'Some\Other\Rest\Controller',
```
so here, I see cases:
- array => scalar (string)
- array => null

maybe it's worth also adding:
- scalar (string) => array
- null => array

weierophinney replied:
Actually, we have:

- array + string = array (with string appended)
- null + array = array

I think we need also:

- any type + null == original type
- null + any type == later type
- scalar + array == array (scalar is overwritten)
- scalar + scalar == later scalar (original scalar is overwritten)

I'll get those test cases added.

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,194 @@
+    /**
+     * Merge two arrays together.
+     *
+     * If an integer key exists in both arrays and preserveNumericKeys is false, the value
```
description should be updated, as we do not have `preserveNumericKeys` property anymore.

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,194 @@
+                continue;
+            }
+
+            $value = $this->replace($value, $newKey);
```
we have the same call inside `if ` above, we can move it before the statement and use inside the value, as we are replacing it anyway.

michalbundyra commented on test/TestAsset/ConfigPostProcessor/MergeEquivalentKeys.php:
```diff
@@ -0,0 +1,16 @@
+            'name' => 'Some\Rest\Controller',
+        ],
+    ],
+    'zend-expressive' => null,
```
I am a bit confused here, because on the result it looks like it's omitted. You haven't documented that case, but I believe it's fine? So it is not possible to nullify the configuration, right?

weierophinney replied:
Correct. This is more explicit in the changes I'm about to push.

michalbundyra reviewed (commented):
@weierophinney 

> The code is already following the same merge algorithm as zend-config-aggregator, which means:
> 
> - If the first value is a scalar, the second value overwrites.
> - If the first value is an array, and the second value is scalar, the second value is appended.
> - If both values are arrays, they are merged, with the second taking precedence.

`null` is not a scalar. What should happen if:
- the first value is null and 2nd is not
- the 2nd value is null and the first is not

In test, what I noted in comments, I am missing also cases:
- scalar => array
- null => array


In general changes looks good. 

I am thinking only if replacing everything in the configuration is the way we should go.
On migration is a bit different story, because this is just a tool which help update application to Laminas.
The tool which is replacing whole configuration of 3rd part libraries is quite risky: for example we can have routing defined in 3rd part library which can be changed, but links can be used separately on frontend (in the same 3rd party library) and these will be not changed.
A bit safer solution will be to replace only:
- configuration keys - so first level keys (but using just part of the rules - just configuration keys)
- service manager/plugin managers configuration (I guess including custom plugin managers)

I think we shouldn't touch routing completely.
Also something what is not Laminas/Expressive/Apigility configuration on 1st level - so custom 3rd party configuration.

It looks like we might have an issue also with `invokable` configuration in SM (this is array `alias => target` or if the alias is not provided target is registered).
What we would need do here is:
- if alias is provided do not rewrite it
- otherwise (alias is not provided) - add class alias as the target is, and rewrite the target.

I'll have another look in the morning. For now it looks that in most of cases it's gonna work fine. By rewriting everything (recursion, without checking what was the 1st level key) it is similar to approach we take first with migration script.

weierophinney commented:
> I think we shouldn't touch routing completely.

I disagree, and specifically because of Apigility.

Apigility defines a number of routes, and the top-level name of the route is "zf-apigility". If we do not touch that, we potentially have some issues if folks are adding items to it (in particular, documentation paths).

But this could potentially be considered a special case, and we could alter _only_ that one.

weierophinney committed:
fix: add zf-apigility => apigility replacement

As an exact replacement, as the zfcampus/zf-apigility package defines
the top-level key.

weierophinney committed:
fix: correct typo in comment

weierophinney committed:
docs: fix description of ConfigPostProcessor::merge

Removes verbiage around `$preserveNumericKeys`, which is no longer
relevant.

weierophinney committed:
qa: reuse return value from replace

Instead of calling `$this->replace($value, $newKey)` in multiple
locations, call it once, and re-use it withing conditionals.

michalbundyra commented:
@weierophinney 
> Apigility defines a number of routes, and the top-level name of the route is "zf-apigility".

That's correct, I've checked it and for example:
https://github.com/zfcampus/zf-apigility-documentation/blob/master/config/module.config.php#L15

But the problem with this is that when you build the route you need provide the key to, so if 3rd-party library adds some route it will be unusable in the views because we are not rewriting views. So in views it will be still something like `$router->getPath('zf-apigility/documentation/something')`... 

It looks like trying to add Laminas compatibility for everything is going to be impossible.

weierophinney committed:
fix: ensure rewrite rules are as expected

Adds test cases for:

- Merging null with scalar (scalar overwrites)
- Merging scalar with null (does not overwrite)
- Merging array with null (no change to array)
- Merging scalar with array (array overwrites)
- Merging two scalars (second overwrites first)

Due to these tests, I identified code changes necessary to make the post
processor behave per the expectations.

weierophinney committed:
qa: simplify conditional

Use isset instead of array_key_exists for check.

weierophinney commented:
> It looks like trying to add Laminas compatibility for everything is going to be impossible.

I never expected we'd get 100% compatibility! The fact that we're as far as we are is mind-blowing, as this configuration post processing piece answers essentially all but one piece of open feedback!

With regards to the routing bit: I think you're right, in the end. The only routes we actually ship that are problematic are the ones in Apigility — and there's very little chance that folks are extending those routes, as they're **specific to the admin UI**. Even the documentation one is unlikely to be an issue, as those are handled by our own modules; if they're handled by somebody else's, they're likely using their own routing table. If they are not, that becomes an FAQ item.

My current thinking at this point is:

- We rewrite as much as possible in configuration, recursively, with a few exceptions/special cases (e.g., aliases, invokables, routing). HOWEVER,
- Any time we rewrite a **key**, we keep the original key with unchanged contents as well, side-by-side with the altered version. While this expands the configuration, it also means developers can compare the two at runtime.

This would be a trivial change to what I'm already doing, and I think would alleviate most of your own concerns.

weierophinney committed:
feat: ignore routing configuration

Since URL generation is dependent on the route names, we should not
rewrite them.

weierophinney committed:
feat: add support for processing invokables

This patch does two things.

First, it modifies how replacements are matched and triggered, passing a
hierarchical array of keys so that rulesets can determine where they are
in the hierarchy, as well as access the keys present. This allows us to,
for instance, know when we are matching a top-level key.

Second, it adds a ruleset for rewriting invokables configuration. If the
alias does not change but the target does, then that change is made. If
the alias changes, too, then:

- An entry is added for the original alias, pointing to the rewritten target.
- An entry is added for the new alias, pointing to the rewritten target.

This ensures aliases work as expected, and the original service names
are still present in configuration.

weierophinney committed:
feat: keep original top-level keys and contents

We want to keep original top-level keys and contents so that developers
can compare the two, and so that any factories that depend on the old
keys will continue to work.

weierophinney commented:
**Updated**: I updated the content to incorporate the changes introduced in a22283c

@michalbundyra Okay, changes are all created and pushed!

Summary of changes since last review:

- Fixed various typos and removed extraneous docs
- Added "zf-apigility" => "apigility" as an explicit key change
- Calls `$this->replace()` once for the value, and re-uses the result
- Adds tests for the following rewrite merge precedence rules, and fixes code to conform:
  - Merging null with scalar (scalar overwrites)
  - Merging sclar with null (does not overwrite; scalar persists)
  - Merging array with null (does not overwrite; array persists)
  - Merging scalar with array (array overwrites)
  - Merging two scalars (last wins)
- Added ruleset to ignore "router" (occurring at top-level) configuration
- Added processing of invokables to the ruleset for processing aliases. In such cases, the service name should remain the same, but the target class it creates should be updated. A nice side benefit to this approach is that if the end-user requests the target class name, it will still know how to create it, as those get mapped to the `InvokableFactory`.
- Updated the post processor to keep the original configuration for keys that are rewritten, in parallel with the rewritten key and its contents. This allows both to exist in parallel. This is only done for top-level keys that are renamed; if the key is not renamed, no duplication occurs, and the contents may be renamed.

I think at this point, we should likely merge and tag as soon as we can, so we can get this feature in the hands of users; we can then add more tests, scenarios, and rulesets as we uncover any issues.

weierophinney committed:
fix: better rules for invokables

Invokables are essentially an alias map. This patch modifies the code to
remove the separate logic for rewriting invokables, and instead have the
rewrite rule for aliases also handle invokables.

That way, the rule becomes: rewrite the target only. This ensures the
original service name persists, but now maps to the new target class.

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,265 @@
+     *     nested configuration.
+     * @return array
+     */
+    public function __invoke(array $config, $keys = [])
```
we can have typehint on `$keys` - `array`

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,265 @@
+     * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+     *
+     * @param  array $a
+     * @param  array $b
```
minor: these are redundant as we have typehints on params

michalbundyra commented on src/ConfigPostProcessor.php:
```diff
@@ -0,0 +1,265 @@
+     * This same logic can be used for invokables, which are essentially just
+     * an alias map.
+     *
+     * @param array $value
```
invalid - as we do not have this param anymore in this method

michalbundyra commented on src/Module.php:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @param ModuleManager $moduleManager
+     */
+    public function init($moduleManager)
```
I believe we can have typehint here?

weierophinney replied:
I deliberately chose not to so I could test without having a dependency on a package that does not yet exist.

I'll add it once we have transferred the packages to their new homes.

michalbundyra commented on src/Module.php:
```diff
@@ -0,0 +1,53 @@
+     * Type-hinting deliberately omitted to allow unit testing
+     * without dependencies on packages that do not exist yet.
+     *
+     * @param ModuleEvent $moduleEvent
```
invalid param name - inconsistent with signature. Can't we just have typehint on param?

weierophinney replied:
I'll update the annotation; see my comment about the `init()` method for why I don't have an actual typehint here.

michalbundyra commented on test/ConfigPostProcessorTest.php:
```diff
@@ -0,0 +1,43 @@
+class ConfigPostProcessorTest extends TestCase
+{
+    /**
+     * @return iterable
```
I am not sure if it is right type, always thought it should be Generator for yield

weierophinney replied:
`iterable` works for any value that can be iterated. That includes generators, iterators, and arrays. I chose `iterable` so that we can refactor this to use an array later if we want.

michalbundyra reviewed (approved):
@weierophinney It looks fine now, just some very minor comments.

weierophinney committed:
qa: incorporate feedback from review

- Removes unnecessary parameter annotations.
- Adds typehints where needed and useful.
- Fixes annotations where they were incorrect.
